### ISSUE TITLE
Stack overflow warning when array is too large.

### ISSUE THREAD

**Issue by sharifhsn** [rust-lang/rust-clippy#9042]
### What it does

Arrays always allocate on the stack, not on the heap. If an array is declared which is too large for the stack, it will cause a stack overflow at runtime. This can be confusing behavior because it is an unexpected abort in an opaque operation.

This lint detects when the size of an array will/is likely to overflow the stack and warn the user that this type is likely to overflow the stack. I believe this limit is 2 MB generally, depending on the architecture and operating system.

The lint recommends an appropriate `Vec` to be used instead. I used the `vec!` macro in my example but I'm unsure if this would work in all cases. For example, if a type annotation is used, that annotation would also have to be changed/erased.

I think the idea behind this lint could be extended as well, to other types that can stack overflow based on parameters. I'm not sure how this would interact with const generics, but I think if this idea could work on them that would also be good.

Example lint text:
"This array is too large and is likely to trigger a stack overflow when run. It should be allocated on the heap in a `Vec` instead."

### Lint Name

array_too_big

### Category

suspicious

### Advantage

- prevents stack overflows that are commonly caused by newbies (like me) who try to make massive multi-dimensional arrays

### Drawbacks

- could possibly be overly conservative in an environment with a much larger stack than usual where heap allocations are more of a concern (embedded?)

### Example

```rust
fn main() {
    let arr = [0u64; 0x100000]; // This array attempts to allocate 8 MB on the stack!
    println!("{arr:?}");
}
```

Could be written as:

```rust
fn main() {
    let arr = vec![0u64; 0x100000];
    println!("{arr:?}");
}
```

**Comment by sharifhsn** [comment]
I just checked lints again, and the `large-stack-arrays` lint already does the same thing.
